File: prototype.php
<?php
    $db = new SQLite3('/tmp/mysqlitedb.db');
    //print "hello";
    $table_exists = $db->query("SELECT COUNT(*) as count FROM sqlite_master WHERE type='table' AND name='".$db->escapeString($_GET['tube'])."' LIMIT 1");
    $table_exists = ($table_exists->fetchArray(SQLITE3_ASSOC)['count'] == 0) ? False : True;
    //var_dump($table_exists);
    //var_dump($_SERVER['REQUEST_METHOD']);
    switch($_SERVER['REQUEST_METHOD'])    {
        case 'POST':
            if($table_exists == False)    {
                $db->exec("CREATE TABLE IF NOT EXISTS ".$db->escapeString($_GET['tube'])." (ID INTEGER PRIMARY KEY AUTOINCREMENT,value TEXT NOT NULL,state INTEGER NOT NULL,ts INTEGER,ttr INTERGER)");            
            }
            $write_conform = $db->exec("INSERT INTO ".$db->escapeString($_GET['tube'])." (value,state,ts) VALUES ('".$_POST['task']."', 0, ".time().");");
            if($write_conform == true)  {
                print json_encode(array('tube' => $_GET['tube'], 'id' => $db->lastInsertRowID()));
            } else {
                print json_encode(array('tube' => $_GET['tube'], 'id' => false));
            }
            break;
        case 'GET':
            if($table_exists == False)    {
                print json_encode(array('tube' => $_GET['tube'], 'tasks' => False));
            } else {
                //$count = (is_int($_GET['count']) && $_GET['count'] > 0) ? $_GET['count'] : 1;
                $db->exec("BEGIN EXCLUSIVE TRANSACTION");
                $task_q = $db->query("SELECT ID as id,value FROM ".$db->escapeString($_GET['tube'])." WHERE state=0 ORDER BY ID ASC LIMIT 1")->fetchArray(SQLITE3_ASSOC);
                if($task_q != False){
                    $ts = time();
                    $db->exec("UPDATE ".$db->escapeString($_GET['tube'])." SET state=1,ts=".$ts." WHERE ID=".$db->escapeString($task_q['id']));
                } else {
                    $task_q = array();
                }
                $db->exec("COMMIT");
                
                print json_encode(array('tube' => $_GET['tube'], 'tasks' => $task_q, 'ts' => $ts));
            }
            break;
        case 'DELETE':
            if($table_exists == True)    {
                if($_GET['ts']) {
                    $d_result = $db->exec("DELETE FROM ".$db->escapeString($_GET['tube'])." WHERE ID=".$db->escapeString($_GET['id'])." AND ts=".$db->escapeString($_GET['ts']));
                } else {
                    $d_result = $db->exec("DELETE FROM ".$db->escapeString($_GET['tube'])." WHERE ID=".$db->escapeString($_GET['id']));
                }
                if($d_result == True && $db->changes() > 0)   {
                    print json_encode(array('tube' => $_GET['tube'], 'deleted' => True));
                } else {
                    print json_encode(array('tube' => $_GET['tube'], 'deleted' => False));
                }
            }
            break;
    }

File: prototype.py
import tornado.ioloop
import tornado.web
from tornado.escape import json_encode

import sqlite3
import time

class queueDB():        
    def setQueue(self, queue):
        self.db = sqlite3.connect('/tmp/'+queue+'.db')
        self.db.row_factory = self.dict_factory
        
    def tubeCreate(self, force=None):
        self.db.execute("CREATE TABLE IF NOT EXISTS tube (ID INTEGER PRIMARY KEY AUTOINCREMENT,task TEXT NOT NULL,state INTEGER NOT NULL,ts INTEGER,ttr INTERGER)")
        
    def tubeExists(self):
        c = self.db.cursor()
        c.execute("SELECT COUNT(*) as count FROM sqlite_master WHERE type='table' AND name='tube' LIMIT 1");
        if c.fetchone()['count'] == 1:
            return True
        else:
            return False
    
    def tubeAddTask(self, task, ttr=60):
        c = self.db.cursor()
        c.execute("INSERT INTO tube (task,state,ts,ttr) VALUES (?, 0, ?, ?);", (task, time.time(), ttr,))
        self.db.commit()
        if c.rowcount > 0:
            return c.lastrowid
        else:
            return False
    def tubeGetTask(self, Id=None, ttr=None):
        c = self.db.cursor()
        c.execute("BEGIN EXCLUSIVE TRANSACTION");
        c.execute("SELECT ID as id,task FROM tube WHERE state=0 OR (state=1 AND ts+ttr < ? ) ORDER BY ID ASC LIMIT 1", (time.time(),))
        task = c.fetchone()
        if task != None:
            ts = time.time();
            if ttr == None:
                c.execute("UPDATE tube SET state=1,ts=? WHERE ID=?", (ts, task['id']))
            else:
                c.execute("UPDATE tube SET state=1,ts=?,ttr=? WHERE ID=?", (ts,ttr, task['id']))
            task['ts'] = ts
        else:
            task = None
        
        self.db.commit()
        return task
    def tubeTouchTask(self, Id, timestamp):
        """
        update the timestamp of an open task
        if a user has taken a task and they need more time before ttr
        they should use this to buy more time
        """
        c = self.db.cursor()
        ts = time.time()
        c.excute("UPDATE tube SET ts=? WHERE ID=? AND state=1 AND ts=?", (ts,Id,timestamp,))
        self.db.commit()
        if c.rowcount > 0:
            return ts
        else:
            return false
    
    def tubeRmTask(self, Id, ts=None):
        """
        Delete a task using its Id, if the timestamp
        is provded it will check with that
        """
        c = self.db.cursor()
        if ts==None:
            c.execute("DELETE FROM tube WHERE ID=?", (Id,))
        else:
            c.execute("DELETE FROM tube WHERE ID=? AND ts=?", (Id,ts,))
        self.db.commit()
        
        if c.rowcount > 0:
            return True
        else:
            return False
        
    def dict_factory(self, cursor, row):
        d = {}
        for idx, col in enumerate(cursor.description):
            d[col[0]] = row[idx]
        return d

class MainHandler(tornado.web.RequestHandler):
    def get(self):
        self.write("Hello, world")
        
class taskGet(tornado.web.RequestHandler, queueDB):
    def get(self, queue, task=None):
        self.setQueue(queue)
        ttr = self.get_argument("ttr",None)
        task = self.tubeGetTask(ttr=ttr)
        if task != None:
            self.write(json_encode({'tube' : queue, 'task': task['task'], 'id':task['id'], 'ts':task['ts']}))
        else:
            self.write(json_encode({'tube' : queue, 'task': None, 'id':None, 'error' : 'No task avaiable'}))

class taskAdd(tornado.web.RequestHandler, queueDB):
    def put(self, queue):
        try:
            self.setQueue(queue)
            
            if self.tubeExists() == False:
                self.tubeCreate()
            ttr = self.get_argument("ttr",60)
            task_id = self.tubeAddTask(self.get_argument('task'), ttr=ttr)
            if task_id != False:
                self.write(json_encode({'tube' : queue, 'id': task_id}))
            else:
                self.write(json_encode({'tube' : queue, 'id': False}))
                
        except tornado.web.MissingArgumentError:
            self.write(json_encode({'tube' : queue, 'error': 'No task provided'}))
        
        self.db.close()

class taskTouch(tornado.web.RequestHandler, queueDB):
    def get(self):
        self.write("Hello, world")
        
class taskRm(tornado.web.RequestHandler, queueDB):
    def delete(self, queue, Id):
        self.setQueue(queue)
        rm = self.tubeRmTask(Id, self.get_argument("ts",None))
        self.db.close()
        self.write(json_encode({'tube' : queue, 'id': Id, 'deleted':rm}))
        
application = tornado.web.Application([
    (r"/queue/(.*)/task/get", taskGet), #get the next task
    (r"/queue/(.*)/task/get/(.*)", taskGet), #get an exact task
    (r"/queue/(.*)/task/add", taskAdd),
    (r"/queue/(.*)/task/touch", MainHandler),
    (r"/queue/(.*)/task/rm/(.*)", taskRm),
    (r"/task/peak", MainHandler),
    (r"/task/free", MainHandler),
    (r"/task/bury", MainHandler),
    (r"/task/kick", MainHandler),
])

if __name__ == "__main__":
    application.listen(8888)
    tornado.ioloop.IOLoop.instance().start()
